File: graphs/cheapest_flights_within_k_stops.rb
# @param {Integer} n
# @param {Integer[][]} flights
# @param {Integer} src
# @param {Integer} dst
# @param {Integer} k
# @return {Integer}
def find_cheapest_price(n, flights, src, dst, num_of_stops_allowed)
    prev_cost = [1.0/0.0]*n
    prev_cost[src] = 0
    curr_cost = prev_cost.dup
    for i in 0..num_of_stops_allowed
        flights.each do |src1, src2, cost|
            if (prev_cost[src1] != 1.0/0.0)
                curr_cost[src2] = [prev_cost[src1] + cost, curr_cost[src2]].min
            end
        end
        prev_cost = curr_cost.dup
    end
    return prev_cost[dst] != 1.0 / 0.0 ? prev_cost[dst] : -1
end

File: graphs/cheapest_flights_within_k_stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        prev_cost = [float('inf')] * n
        prev_cost[src] = 0

        curr_cost = prev_cost.copy()

        for i in range(k + 1):
            for src1, src2, cost in flights:
                if prev_cost[src1] != float('inf'):
                    curr_cost[src2] = min(prev_cost[src1] + cost, curr_cost[src2])

            prev_cost = curr_cost.copy()

        return prev_cost[dst] if prev_cost[dst] != float('inf') else -1